fluentd/lib/fluentd-okd-config-generator/lib/fluentd-okd-config-generator/generator.rb

```diff
@@ -0,0 +1,141 @@
+module FluentdOKDConfigGenerator
+    
+    #
+    # Generate the 'match' blocks for the @OUTPUT label
+    # to split source messages to source labels
+    #
+    class OutputLabelMatches
+        include ConfHelpers
+        def initialize(source, tags)
+            @template_name = 'output_label_match.erb'
+            @label_name = label_name(source)
+            @tags = tags
+        end
+    end
+    
+    #
+    # Generate the source labels to match and split for each target label
+    #
+    class SourceLabelConf
+        include ConfHelpers
+        def initialize(source, targets)
+            @template_name = 'source_label.erb'
+            @source = source
+            @targets = targets
+            @label_name = label_name(source)
+        end
+        
+        def relabel(target)
+            "#{@label_name}_#{target[:type].upcase()}#{target[:counter]}"
+        end
+    end
+    
+    #
+    # Generate the target label blocks for each endpoint
+    #
+    class EndpointLabelConf
+
+        include ConfHelpers
+
+        def initialize(source, target)
+            @template_name = 'endpoint_label.erb'
+            @source = source
+            @target = target
+            set_and_format_label_name(source, target)
+            set_and_format_retry_tag(source)
+        end
+
+        def store_id(target, context)
+            store_id = "#{target[:store_id]}#{target[:counter]}"
+            context == :prefix_as_retry ? "retry_#{store_id}" : store_id
+        end
+
+        def buffer_path(target, context)
+            "/var/lib/fluentd/#{store_id(target,context)}"
+        end
+
+        def set_and_format_label_name(source, target)
+            @label_name = "@#{source.split('.').join('_').upcase()}_#{target[:type].upcase()}#{target[:counter]}"
+        end
+        
+        def set_and_format_retry_tag(source)
+            @retry_tag = "retry_#{source.split('.').join('_').downcase()}"
+        end
+    end
+
+    class Generator
+        def initialize(logger)
+            @log = logger
+        end
+
+        def generate(input)
+            @log.debug("Generating config for input: #{input}")
+            results = input.collect do | source, config |
+                gen_output_conf(source, config[:targets])
+            end
+            @log.debug("Results: #{results}")
+            "".tap do | buffer |
+                results.each do |conf|
+                    buffer << "\n"
+                    conf.each do | k, value|
+                        buffer << value.join('\n')
+                    end
+                end 
+            end
+        end
+        
+        def gen_output_label_match_confs(sources, tags)
```

Comment:
What does this function do?  Are you trying to differentiate which outputs should be the default/fallthrough/last options vs. those which are not?  @nhosoi how was this done in the rsyslog linux system role?

Reply:
This generates the match configs which relabel tags to labels and are included at the end of the @INGRESS label.  The flow is diagramed in the design doc

Reply:
> @nhosoi how was this done in the rsyslog linux system role?

This is an example of vars.yaml.  In this file, depending on the log collections, it switches the output.
https://github.com/linux-system-roles/logging/#varsyaml

Currently, there's only one output_role elasticsearch is defined.  But we could add other output roles there.
$ ls ~/linux-system-roles/logging/roles/rsyslog/roles/output_roles
elasticsearch/